### ISSUE TITLE
Nextcloud Talk is really slow when using HTTP2

### ISSUE THREAD

**Issue by sL5gqj6Q** [nextcloud/spreed#2211]
## Steps to reproduce
1. Install Nextcloud Talk App from the appstore in the browser
2. User A starts a chat with User B, both registered on the same Nextcloud server
3. Messages take up to 20 seconds to be delivered
4. Also replies from User B to User A need up to 20 seconds to be delivered

### Expected behaviour
I expect messages to be delivered nearly instantaneuously

### Actual behaviour
Nextcloud Talk feels slow and laggy. User A opens a chat with User B in his Browser. There is a constant loading circle around User A, as if he cannot join the session. User B is invited. Chat messages from both users take up to 20 seconds to send/receive.

## Browser

**Microphone available:** yes

**Camera available:** yes

**Operating system:** User A: Linux Mint 19.2 , User B: Android 9

**Browser name:** User A: Firefox 69 , User B: Nextcloud Talk Android App v6.0.0.beta3

**Browser version:** see above

## Talk app

**Talk app version:** 5.0.4

**Custom TURN server configured:** no

**Custom STUN server configured:**no

## Server configuration

**Operating system**: Raspbian 8.0 (jessie)

**Web server:** Apache

**Database:** MariaDB

**PHP version:** 7.0.27

**Nextcloud Version:** 15.0.11

**List of activated apps:**

<details>
Enabled:
  - accessibility: 1.1.0
  - activity: 2.8.2
  - announcementcenter: 3.4.1
  - bruteforcesettings: 1.4.0
  - calendar: 1.6.6
  - checksum: 0.4.3
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 3.1.3
  - dav: 1.8.2
  - federatedfilesharing: 1.5.0
  - federation: 1.5.0
  - files: 1.10.0
  - files_pdfviewer: 1.4.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_versions: 1.8.0
  - files_videoplayer: 1.4.0
  - firstrunwizard: 2.4.0
  - gallery: 18.2.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - music: 0.10.1
  - nextcloud_announcements: 1.4.0
  - notes: 3.0.3
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - password_policy: 1.5.0
  - provisioning_api: 1.5.0
  - quota_warning: 1.4.0
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - spreed: 5.0.4
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - tasks: 0.11.3
  - theming: 1.6.0
  - twofactor_backupcodes: 1.4.1
  - twofactor_totp: 2.1.2
  - updatenotification: 1.5.0
  - workflowengine: 1.5.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_ldap

</details>

**Nextcloud configuration:**

<details>
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "***REMOVED URL***",
            "***REMOVED URL***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "http:\/\/localhost\/nextcloud",
        "dbtype": "mysql",
        "version": "15.0.11.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "updater.release.channel": "stable",
        "maintenance": false,
        "theme": "",
        "loglevel": 2,
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "tls",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "memcache.local": "\\OC\\Memcache\\APCu"
    }
}

</details>

### Server log (data/nextcloud.log)
<details>
not found
</details>

**Comment by tiotrom** [comment]
I can confirm it is the same for me. We use a VPS  with 4GB of RAM and 8 core for nextcloud. Tested with Firefox and Android app. It is always very slow. I was thinking it might be our setup. We also have a TURN server in place.

**Comment by nickvergessen** [comment]
Are you using HTTP2?

**Comment by sL5gqj6Q** [comment]
Hi @nickvergessen ,
yes, my Nextcloud server supports HTTP2 .

**Comment by nickvergessen** [comment]
Http2 Support is currently a problem with chatting and webrtc

**Comment by tiotrom** [comment]
In my case I don't use HTTP2 and the chat is quite slow as well. But not 20 seconds. More about 2 seconds. Ok, I don't want to interrupt the conversation but I thought it may be useful. Talk is a bit slow - not "instant".

**Comment by nickvergessen** [comment]
@sL5gqj6Q so try to disable HTTP2 for now, until we found a useful working solution to the problem. Sorry about your inconvenience.

@tiotrom 2 seconds is not related to this problem here. But you still do not need to create a new issue for it. We wait by default 1 second to reduce the load for the server. Otherwise every user would bombard the server with requests all the time.

**Comment by sL5gqj6Q** [comment]
After disabling http2, messaging is possible nearly in real time. Thanks for the hint @nickvergessen . However, I hope that http2 support is on the agenda as one of the features to add to Nextcloud Talk.

**Comment by nickvergessen** [comment]
So after a discussion with a college today, I tried it again and for me it worked quite well.
I followed the following manual to set up http2: https://techwombat.com/enable-http2-apache-ubuntu-16-04/

maybe you forgot one of the changes on your system?

**Comment by cjhille** [comment]
I'm running Nextcloud behind jwilders proxy and even manually disabling http2 (and reloading the proxy config) didn't solve the described issues for me. Is there any other way to make sure http2 is not used? Apparently http2 can't be passend with nginx reverse proxies properly :(

The reverse proxy setup is officially supported/maintained in the nextcloud docs (examples folder) so I'd suggest to add the info, that spreed won't be working properly behind reverse proxies.

**Comment by warnerbryce** [comment]
Even without HTTP2 disabled, i have a real slow down on Talk with Nc 17.0.2 5-6 seconds before the text is send.

**Comment by jaabrao** [comment]
I was dealing with the issue of Talk being really slow, and stumbled upon this page after some searching. I see some people pointing at HTTP2 being enabled as the culprit.

Alas, when I checked, it was never on to begin with on my current server, nor on the old one.

So I decided to enable it. And following all the instructions still left me with http1.1. The following link shared by nickvergessen nailed it:

> I followed the following manual to set up http2: https://techwombat.com/enable-http2-apache-ubuntu-16-04/

After doing that - and making sure it tests positive for HTTP2 operation - it's fast as it should be.

**Comment by Bellusterra** [comment]
it seems to have become more of an issue after the most recent update of Nextcloud Talk. but that could just be me.

**Comment by cjhille** [comment]
I checked the logs because my entire NC instance slowed down when switching back and forth between channels quickly. I noticed this in my container log:

> *35272 upstream server temporarily disabled while reading response header from upstream, client: 109.104.41.102, server: cloud.REDACTED.de, request: "DELETE /ocs/v2.php/apps/spreed/api/v1/room/u3i7skpi/participants/active HTTP/2.0", upstream: "http://172.19.0.7:80/ocs/v2.php/apps/spreed/api/v1/room/u3i7skpi/participants/active", host: "cloud.REDACTED.de"

> *35272 no live upstreams while connecting to upstream, client: 109.104.41.102, server: cloud.REDACTED.de, request: "DELETE /ocs/v2.php/apps/spreed/api/v1/room/u3i7skpi/participants/active HTTP/2.0", upstream: "http://cloud.REDACTED.de/ocs/v2.php/apps/spreed/api/v1/room/u3i7skpi/participants/active", host: "cloud.REDACTED.de"

This explains why the entire NC instance freezes. NGINX temporarily disables the upstream. @nickvergessen Could it be that *WITHOUT* HTTP2 :
* some responses take too long (some sort of long polling?) -> connection gets terminated with error -> nginx temporarily disables entire upstream?
* talk makes too many concurrent requests and some capacity limit is reached?

**Comment by Bellusterra** [comment]
@cjhille you are running nextcloud docker behind nginx reverse proxy?

**Comment by cjhille** [comment]
Yes, basically I'm using [this official docker config](https://github.com/nextcloud/docker/tree/master/.examples/docker-compose/with-nginx-proxy/mariadb-cron-redis/fpm) with an additional reverse proxy ([jwilder/nginx-proxy](https://github.com/jwilder/nginx-proxy)) in front for certificate automation.
However there are two sequential upstreams:
1. from jwilder/nginx to nextcloud nginx
2. from nextcloud nginx to php-fpm

The previous error log was from the first upstream, however the latter seems to be causing the issues for the former.

Error from the second (nginx to php-fpm) upstream:
> [error] 6#6: *44159 upstream timed out (110: Operation timed out) while reading response header from upstream, client: 109.104.41.102, server: , request: "GET /ocs/v2.php/apps/spreed/api/v1/room/gg6g7z3t/participants HTTP/1.1", upstream: "fastcgi://172.18.0.9:9000", host: "cloud.REDACTED.de"

**Comment by Bellusterra** [comment]
@cjhille I have a similar setup, nginx reverse proxy to another nginx in front of the docker nextcloud apache version. I am not getting the same errors you are however.  Both nginx are run directly on the host machine however and not in docker, if that helps with troubleshooting at all.

**Comment by Bellusterra** [comment]
This may help you as well. I don't remember where I got these settings, but you can always give them a try and see if they fix your error and if not remove and restart your nginx.

http {                                                                                         
                                                                                              
# INCREASE NGINX PROXY TIME-OUT                                                                

proxy_connect_timeout 600;
proxy_send_timeout 600;
proxy_read_timeout 600;
send_timeout 600;

**Comment by Dennis1993** [comment]
My nextcloud instance is slowly too. The messages will be delivered after 30 seconds and more.

PHP 7.4
MySQL 5.7
NextCloud 18.0.1
HTTP/2 activated

On a second server the same problem (PHP 7.3, MySQL 5.7, HTTP/1.1). A video call needs ober 1 minute to establish a connection to the partner.

**Comment by nickvergessen** [comment]
You forgot to post your webserver version + configuration.
That is the most troublesome point in this discussion.

**Comment by nickvergessen** [comment]
So I checked this again because of all the comments. and it works very fine here:

```
$ apache2 -v
Server version: Apache/2.4.41 (Ubuntu)
Server built:   2019-08-21T20:43:38

$ php --version
PHP 7.3.15-1+ubuntu19.10.1+deb.sury.org+1 (cli) (built: Feb 20 2020 12:24:05) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.15, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.15-1+ubuntu19.10.1+deb.sury.org+1, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.9.2, Copyright (c) 2002-2020, by Derick Rethans
```

To enable it I did:
```
sudo a2dismod php7.3
sudo a2enconf php7.3-fpm
sudo a2dismod mpm_prefork
sudo a2enmod mpm_event
sudo a2enconf php7.3-fpm 
```

**Comment by cjhille** [comment]
@nickvergessen I'm glad that see that you have a specific configuration (custom apache+php-fpm) that works flawlessly in your testing. However I'm quite sure, that these days many people prefer just to use the official docker images, which includes a nginx+php-fpm/alpine variant (as they are demonstrated in the [official docs](https://github.com/nextcloud/docker/tree/master/.examples/docker-compose/with-nginx-proxy/mariadb-cron-redis/fpm)). Have you tested running talk behind nginx? I'm guessing the underlying problem might be a configuration issue with talks long running requests.

Out of curiosity: If talk requires some more dependencies (i.e. the php modules you mentioned), how are you suggest making sure, that they are contained (or can be auto installed) in official images? Shouldn't Talk be trying to support the official images?

**Comment by warnerbryce** [comment]
I would say exactly the same. 
The conf have to work on Apache and Nginx, both are hugely used in Nextcloud.

**Comment by nickvergessen** [comment]
Hope this doesn't sound rude, at least it's not meant to be!

No, I have not tested nginx. Sorry to say, but it's not officially supported by Nextcloud GmbH:
https://docs.nextcloud.com/server/18/admin_manual/installation/nginx.html
> These configurations examples were originally provided by \@josh4trunks and are community-maintained.

Also the vm, snap, … are not officially supported:
https://docs.nextcloud.com/server/18/admin_manual/installation/source_installation.html
> Please note that those two options are not officially supported by Nextcloud GmbH.

It's simply impossible for us with our teamsize to do everything.
Nextcloud Talk works with HTTP2 on Apache2, so it is not a software related problem in the code of Nextcloud Talk, but a configuration issue with your nginx server.
You can either ask the Docker people to have a look (if they know what is causing it), or you try to find the josh4trunks user and ask him if he has an idea why nginx is not capable to manage more then 1 connection per window with HTTP2

> Out of curiosity: If talk requires some more dependencies (i.e. the php modules you mentioned), how are you making sure, that theses deps are contained (or can be auto installed) in official images? Shouldn't Talk be trying to support the official images?

There are no such dependencies for talk, that is what you need on apache2 for http2

**Comment by Kovo** [comment]
Same issue here on a fresh install of NC using nginx. Did not use the docker image. But installed it manually. Two people on a call works fine for the most part. Until a third joins, and then everything bogs down to a near halt. Even after ending the call, its still slow until it eventually fixes itself.

**Comment by nickvergessen** [comment]
> I can also confirm that there is no difference between http2/http1.1

Then you are having a very different issue

**Comment by Kovo** [comment]
> > I can also confirm that there is no difference between http2/http1.1
> 
> Then you are having a very different issue

How so? People in this thread have stated issues both with and without http2. Symptoms are also almost identical.

**Comment by JohnArcher** [comment]
We are running Nextcloud 18.0.3 and Talk 8.0.5/8.0.7 (tested both version). Our instance runs on the shared hoster all-inkl.com (Package "all-inkl Business") which is relatively popular here in Germany. I don't expect speed wonders, but also suffer from this 20 to 30 seconds lag for every message and permanent loading circles next to participant names.

We run on a Apache webserver (don't see the version number, but I assume it is quite the latest) and this infos from phpinfo:
`
System: Linux ####### 4.15.0-91-generic #92~16.04.1-Ubuntu SMP Fri Feb 28 14:57:22 UTC 2020 x86_64

Server API: FPM/FastCGI

PHP Version: 7.3.11-nmm1

This program makes use of the Zend Scripting Language Engine:
Zend Engine v3.3.11, Copyright (c) 1998-2018 Zend Technologies
    with the ionCube PHP Loader + ionCube24 v10.3.9, Copyright (c) 2002-2019, by ionCube Ltd.
    with Zend OPcache v7.3.11-nmm1, Copyright (c) 1999-2018, by Zend Technologies

$_SERVER['HTTP2']: on

$_SERVER['SERVER_PROTOCOL']: HTTP/2.0
`
Is there anything else I could provide from phpinfo or do you have an idea or experience with all-inkl.com what to do in order to increase the speed? Thanks in advance!

**Comment by alve89** [comment]
I'm also on a shared host of all-inkl.com (all-inkl Premium) without having HTTP2 activated (yet). At all-inkl.com it's very easy to switch to a machine with HTTP2, so if the problem of [@JohnArcher is solved I could move there.

Or why is it even on HTTP 1.1 this slow?

Chat messages: 20+ sec
Calls: Nearly impossible (I didn't manage to run a single successful call yet)

**Comment by nickvergessen** [comment]
Can you share a public conversation link with me or a test account on your instance?

**Comment by alve89** [comment]
What would you prefer?

**Comment by alve89** [comment]
I just set a conversation as private one (activating the lobby). When I promote a user as moderator (and downgrade it again) both actions takte action in no time and I got notified about it on all clients. So that worked very quickly.

**Comment by nickvergessen** [comment]
> What would you prefer?

you can do both at the same time `<my github name>@nextcloud.com`

**Comment by alve89** [comment]
You should have mail.

**Comment by JohnArcher** [comment]
> > What would you prefer?
> 
> you can do both at the same time `<my github name>@nextcloud.com`

I have sent you a mail.

**Comment by nickvergessen** [comment]
So, after some checking you have to use:
* php + mpm_prefork or
* php_fpm + mpm_event

All-Inkl uses:
* php_fpm + mpm_prefork

This limits the number of request handlers and causes the effects your are seeing ( https://tecadmin.net/apache-mpm-prefork-and-worker-and-event/ might help to understand).
A mail to support gave the following answer:

> Das wird leider nichts im Webhosting Bereich. Denn die Erweiterung benötigt so viele Ressourcen, dass dies nur auf einen eigenen Server richtig funktionieren wird.
>
> ---
> Google translate:
> Unfortunately, this will not work in the web hosting area. Because the expansion needs so many resources that this only work properly on its own server will.

So yeah, Talk won't work on a Webhosting package, but you need a server for it, which totally makes sense.

**Comment by JohnArcher** [comment]
Thanks again for your help yesterday, @nickvergessen !

**Comment by jcastang** [comment]
Hello,

I was facing the same issues:
- Talk slow
- Messages takes up to 20sec to be displayed/sent
- Avatar were loading slow.

I tried with and without HTTP2 -> No differences.
I am on Ubuntu 18.04 and Nextcloud 18.0.3 / Talk 8.0.6

I was running with mpm_event and php-fpm.
After going back to mod_php and mpm_prefork all issues were gone.

Please test this configuration and feel free to report :)

Cheers,

**Comment by HenryDunham** [comment]
I can confirm: When using NextCloudPi with Nextcloud 18.0.4, Talk 8.0.8, Apache 2.4.38 and PHP 7.3.14, changing to mod_php and mpm_prefork leads to good performance of Talk.

Unfortunately I was not able to get reasonable performance for Talk with http2, php-fpm and mpm_event. I guess it’s an issue related to configuration.

_(Details on what I’ve tried so far: https://help.nextcloud.com/t/nextcloudpi-and-talk-configuration-incompatible/79469)_

**Comment by matzzzl** [comment]
I also made the experience that Nextcloud Talk App v14.1 together with Nextcloud 24.0.3 and the latest Talk app was really slow and laggy so that sending or receiving a simple text message could take up to 20 seconds. After disabling http/2 these performance issues immediately disappeared (all other config options were untouched).

So to me this seems to be an issue which should be solved. Otherwise a configuration tutorial explaining how Nextcloud Talk works properly with http/2 would be helpful.

**Comment by nickvergessen** [comment]
See https://nextcloud-talk.readthedocs.io/en/latest/system-requirements/#webserver

There is nothing talk can fix at the moment.

**Comment by nickvergessen** [comment]
With Talk 15 we will show a dedicated warning in the admin settings about this:
https://github.com/nextcloud/spreed/pull/7516

**Comment by matzzzl** [comment]
I actually used mpm_event. As far as I know, Apache's http2 implementation isn't compatible with mpm_prefork at all.

And within php fpm I also set pm.max_children = 64.

**Comment by nickvergessen** [comment]
We use this combination (FPM + EVENT) on a daily basis and it works fine.

**Comment by matzzzl** [comment]
I use the php-fpm package (v7.4.3) and the apache2-package (v2.4.41) provided by the Ubuntu 20.04 LTS repositories. Do you have experiences that these versions are incompatible (because, for example, too old)?

**Comment by Chiller2019** [comment]
Hm, will it work with php_fpm + mpm_worker? Because its installed on KeyHelp which use npm_worker for default

**Comment by nickvergessen** [comment]
https://nextcloud-talk.readthedocs.io/en/latest/system-requirements/#webserver

**Comment by matzzzl** [comment]
Finally upgrading to Ubuntu 22.04 (providing php8.1 and apache 2.4.52) solved my performance issues with Nextcloud Talk when http/2 is enabled.

**Comment by matzzzl** [comment]
My PHP modules:

root@xxx:~# php -m
[PHP Modules]
apcu
bcmath
bz2
calendar
Core
ctype
curl
date
dom
exif
FFI
fileinfo
filter
ftp
gd
gettext
gmp
hash
iconv
imagick
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
Phar
posix
readline
Reflection
session
shmop
SimpleXML
sockets
sodium
SPL
standard
sysvmsg
sysvsem
sysvshm
tokenizer
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache


My apache modules:

root@xxx:~# apache2ctl -M
Loaded Modules:
 core_module (static)
 so_module (static)
 watchdog_module (static)
 http_module (static)
 log_config_module (static)
 logio_module (static)
 version_module (static)
 unixd_module (static)
 access_compat_module (shared)
 alias_module (shared)
 auth_basic_module (shared)
 authn_core_module (shared)
 authn_file_module (shared)
 authz_core_module (shared)
 authz_host_module (shared)
 dav_module (shared)
 dav_svn_module (shared)
 authz_svn_module (shared)
 authz_user_module (shared)
 autoindex_module (shared)
 deflate_module (shared)
 dir_module (shared)
 env_module (shared)
 filter_module (shared)
 headers_module (shared)
 http2_module (shared)
 info_module (shared)
 mime_module (shared)
 mpm_event_module (shared)
 negotiation_module (shared)
 proxy_module (shared)
 proxy_fcgi_module (shared)
 reqtimeout_module (shared)
 rewrite_module (shared)
 setenvif_module (shared)
 socache_shmcb_module (shared)
 ssl_module (shared)
 status_module (shared)

If you need additional information, let me know!

**Comment by Grelo4ka** [comment]
I also face this issue with the latest Nextcloud Talk version, installed via NextcloudPi, so it does not seem to be resolved

**Comment by nickvergessen** [comment]
> so it does not seem to be resolved

It's a misconfiguration on your side, nothing we can fix software side.
If you run PHP FPM you need to run the Apache "event" module
FPM + "prefork" will never work

**Comment by Grelo4ka** [comment]
> > so it does not seem to be resolved
> 
> It's a misconfiguration on your side, nothing we can fix software side. If you run PHP FPM you need to run the Apache "event" module FPM + "prefork" will never work

I am running PHP FPM with the Apache event module, as it's the default with NextcloudPi. If the issue can't be fixed due to configuration error, a warning guiding the user on the configuration errors could be added

**Comment by nickvergessen** [comment]
We already do that:

![Image](https://github.com/user-attachments/assets/1db64acb-71d7-4d3a-b15e-f94fe520ab49)

If that does not show for you, would you mind posting the result of `apachectl -V` and the output of phpinfo (has to be done via web, but you can search for `fpm.config` on that page)

**Comment by nickvergessen** [comment]
`Server MPM: event` looks good and is what we are looking for. The other thing we care about is visible on the phpinfo. You can create a `phpinfo.php` file in the root of your Nextcloud server web directory and then call: `https://nextcloud.example.tld/phpinfo.php`
```php
<?php
phpinfo();
```

It looks like this:

![Image](https://github.com/user-attachments/assets/e869b426-29e4-4a2a-9d76-64480b19c133)

The only other thing I can/could imagine is that you simply have too few fpm workers assigned and that is not detectable inside PHP itself. We recommend to have at least the double of the concurrent active participants available, as 1 will be used per person to long poll chat messages (we are working on improving that, but can't tell yet when this is done)

**Comment by Grelo4ka** [comment]
You are right @nickvergessen - increasing `pm.max_children` in `/etc/php/8.3/fpm/pool.d/www.conf` fixed the issue. Thank you!

For reference, if useful:
![Image](https://github.com/user-attachments/assets/de78ea17-6812-4fc0-85aa-78d39eded6ad)
